keel-bakery-plugin/src/main/kotlin/com/netflix/spinnaker/keel/bakery/constraint/BakeConstraintEvaluator.kt

```diff
@@ -0,0 +1,178 @@
+import com.netflix.spinnaker.keel.api.artifacts.DebianArtifact
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.bakery.api.BakeConstraint
+import com.netflix.spinnaker.keel.bakery.api.BakeConstraintAttributes
+import com.netflix.spinnaker.keel.clouddriver.ImageService
+import com.netflix.spinnaker.keel.constraints.ConstraintState
+import com.netflix.spinnaker.keel.constraints.ConstraintStatus.FAIL
+import com.netflix.spinnaker.keel.constraints.ConstraintStatus.PASS
+import com.netflix.spinnaker.keel.constraints.ConstraintStatus.PENDING
+import com.netflix.spinnaker.keel.constraints.StatefulConstraintEvaluator
+import com.netflix.spinnaker.keel.constraints.SupportedConstraintType
+import com.netflix.spinnaker.keel.model.Job
+import com.netflix.spinnaker.keel.orca.OrcaService
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+
+class BakeConstraintEvaluator(
+  repository: KeelRepository,
+  private val imageService: ImageService,
+  private val taskLauncher: TaskLauncher,
+  private val orcaService: OrcaService,
+  private val dynamicConfigService: DynamicConfigService,
+  override val eventPublisher: ApplicationEventPublisher
+) : StatefulConstraintEvaluator<BakeConstraint>(repository) {
+  override val supportedType = SupportedConstraintType<BakeConstraint>("bake")
+
+  override fun canPromote(
+    artifact: DeliveryArtifact,
+    version: String,
+    deliveryConfig: DeliveryConfig,
+    targetEnvironment: Environment,
+    constraint: BakeConstraint,
+    state: ConstraintState
+  ): Boolean {
+    val attributes = state.attributes as BakeConstraintAttributes?
+    if (attributes?.executionId != null) {
+      val execution = runBlocking {
+        orcaService.getOrchestrationExecution(attributes.executionId)
+      }
+      return when {
+        execution.status.isIncomplete() -> {
+          false
+        }
+        execution.status.isFailure() -> {
+          repository.storeConstraintState(state.copy(
+            status = FAIL
+          ))
+          false
+        }
+        else -> {
+          repository.storeConstraintState(state.copy(
+            status = PASS
+          ))
+          true
+        }
+      }
+      // if task failed, mark constraint as failed
+      // if task succeeded just wait for AMI to  appear
+      // if waiting too long time out and fail
+    }
+
+    if (artifact !is DebianArtifact) {
+      repository.storeConstraintState(state.copy(
+        status = PASS
+      ))
+      return true
+    }
+
+    val locatableResources = targetEnvironment
+      .resources
+      .map(Resource<*>::spec)
+      .filterIsInstance<Locatable<*>>()
+
+    val regions: List<String> = locatableResources.flatMap {
```

Comment:
If we're inferring the regions from the resources here (which I think is the right approach), why do you need to specify the regions in the artifact's `vmOptions`?

Reply:
Ultimately we want to get rid of it… baby steps

Reply:
We're going to move in that direction eventually but there are many edge cases. For example, it's not enough just to  consider the immediate environment we're deploying to -- future environments may have different regions but would expect the AMI to already exist. It's also possible the AMI might be used outside the application or outside of the managed parts.

Reply:
Got it. Good point.